es84x/src/main/scala/tech/beshu/ror/es/RorRestChannel.scala

```diff
@@ -17,16 +17,61 @@
 package tech.beshu.ror.es
 
 import org.apache.logging.log4j.scala.Logging
-import org.elasticsearch.rest.{AbstractRestChannel, RestChannel, RestResponse}
+import org.elasticsearch.http.HttpChannel
+import org.elasticsearch.rest.{AbstractRestChannel, RestChannel as EsRestChannel, RestRequest as EsRestRequest, RestResponse as EsRestResponse}
+import squants.information.{Bytes, Information}
+import tech.beshu.ror.accesscontrol.domain.{Address, Header, UriPath}
+import tech.beshu.ror.accesscontrol.request.RequestContext.Method
 import tech.beshu.ror.es.utils.ThreadRepo
+import tech.beshu.ror.syntax.*
+import tech.beshu.ror.utils.RefinedUtils.nes
 
-class RorRestChannel(underlying: RestChannel)
+import java.net.InetSocketAddress
+import scala.jdk.CollectionConverters.*
+
+final class RorRestChannel(underlying: EsRestChannel)
   extends AbstractRestChannel(underlying.request(), true)
     with ResponseFieldsFiltering
     with Logging {
 
-  override def sendResponse(response: RestResponse): Unit = {
+  val restRequest: RorRestRequest = new RorRestRequest(underlying.request())
+
+  override def sendResponse(response: EsRestResponse): Unit = {
     ThreadRepo.removeRestChannel(this)
     underlying.sendResponse(filterRestResponse(response))
   }
 }
+
+final class RorRestRequest(underlying: EsRestRequest) {
+
+  lazy val method: Method = Method.fromStringUnsafe(underlying.method().name())
+
+  lazy val path: UriPath = UriPath
+    .from(underlying.path())
+    .getOrElse(UriPath.from(nes("/")))
+
+  lazy val allHeaders: Set[Header] = Header.fromRawHeaders(
+    underlying
+      .getHeaders.asScala
+      .view.mapValues(_.asScala.toList)
+      .toMap
+  )
+
+  lazy val localAddress: Address =
+    createAddressFrom(_.getLocalAddress)
+      .getOrElse(throw new IllegalArgumentException(s"Cannot create IP or hostname"))
+
+  lazy val remoteAddress: Option[Address] = createAddressFrom(_.getRemoteAddress)
+
+  lazy val content: String = Option(underlying.content()).map(_.utf8ToString()).getOrElse("")
+
+  lazy val contentLength: Information = Bytes(underlying.contentLength())
```

Comment:
WDYT about introducing a trait in the core module for these properties, like BaseRorRestRequest, or sth similar? Then this type can be added to the RequestContext trait, and we can initialize there the request-related data.

Reply:
please take a look (core and es816x module). If it's what you asked, I will port the changes to other es modules

Reply:
LGTM